torch/csrc/jit/passes/utils/memory_dag.cpp

```diff
@@ -93,71 +107,101 @@ bool MemoryDAG::mayContainAlias(
   return all_a_mlocs.intersects(all_b_mlocs);
 }
 
-void MemoryDAG::makePointerTo(Element* from, Element* to) {
-  from->pointsTo.set(to->index);
-  from->cachedMemoryLocations_.clear();
-
-  to->pointedFrom.set(from->index);
+void MemoryDAGBuilder::makePointerTo(Element* from, Element* to) {
+  makePointerToImpl(from, to);
 }
 
-void MemoryDAG::addToContainedElements(Element* elem, Element* container) {
+void MemoryDAGBuilder::addToContainedElements(
+    Element* elem,
+    Element* container) {
   TORCH_INTERNAL_ASSERT(
       elem != container, "Elements cannot contain themselves");
   container->containedElements.set(elem->index);
 }
 
 // Give `v` a fresh alias (i.e. it does not point to any value)
-Element* MemoryDAG::makeFreshValue(const Value* v) {
-  indexToElementMap_.emplace_back(
-      torch::make_unique<Element>(*this, v, indexToElementMap_.size()));
-  return indexToElementMap_.back().get();
+Element* MemoryDAGBuilder::makeFreshValue(const Value* v) {
+  return makeFreshValueImpl(v, indexToElementMap_);
 }
 
-const MemoryLocations& Element::getMemoryLocations() const {
-  if (!cachedMemoryLocations_.empty()) {
-    return cachedMemoryLocations_;
+const MemoryLocations& MemoryDAG::getMemoryLocations(const Element* e) const {
+  if (e->cachedMemoryLocations_) {
```

Comment:
nit: can you add comment to the caching here: when you should expect it to be set